adam_core/orbits/variants.py

```diff
@@ -0,0 +1,177 @@
+        Link variants to the orbits from which they were generated.
+
+        Parameters
+        ----------
+        orbits : `~adam_core.orbits.orbits.Orbits`
+            Orbits from which the variants were generated.
+
+        Returns
+        -------
+        linkage : `~quivr.MultiKeyLinkage[Orbits, VariantOrbits]`
+            Linkage between variants and orbits.
+        """
+        assert orbits.coordinates.time.scale == self.coordinates.time.scale
+
+        # We might want to replace linking on jd1 and jd2 with just linking on mjd
+        # once the changes have been merged
+        return qv.MultiKeyLinkage(
+            orbits,
+            self,
+            left_keys={
+                "orbit_id": orbits.orbit_id,
+                "jd1": orbits.coordinates.time.jd1,
+                "jd2": orbits.coordinates.time.jd2,
+            },
+            right_keys={
+                "orbit_id": self.orbit_id,
+                "jd1": self.coordinates.time.jd1,
+                "jd2": self.coordinates.time.jd2,
+            },
+        )
+
+    def collapse(self, orbits: Orbits) -> Orbits:
+        """
+        Collapse the variants and recalculate the covariance matrix for each
+        each orbit at each epoch. The mean state is taken from the orbits class and
+        is not calculate from the variants.
+
+        Parameters
+        ----------
+        orbits : `~adam_core.orbits.orbits.Orbits`
+            Orbits from which the variants were generated.
+
+        Returns
+        -------
+        collapsed_orbits : `~adam_core.orbits.orbits.Orbits`
+            The collapsed orbits.
+        """
+        link = self.link_to_orbits(orbits)
+
+        # Iterate over the variants and calculate the mean state and covariance matrix
+        # for each orbit at each epoch then create a new orbit with the calculated covariance matrix
+        orbits_list = []
+        for key, orbit, variants in link.iterate():
+            key = key.as_py()
+
+            assert len(orbit) == 1
+
+            samples = variants.coordinates.values
+            mean_state, covariance = mean_and_covariance_from_weighted_samples(
+                samples, variants.weights.to_numpy(), variants.weights_cov.to_numpy()
+            )
+
+            orbit_collapsed = Orbits.from_kwargs(
+                orbit_id=orbit.orbit_id,
+                object_id=orbit.object_id,
+                coordinates=CartesianCoordinates.from_kwargs(
+                    time=orbit.coordinates.time,
+                    x=orbit.coordinates.x,
+                    y=orbit.coordinates.y,
+                    z=orbit.coordinates.z,
+                    vx=orbit.coordinates.vx,
+                    vy=orbit.coordinates.vy,
+                    vz=orbit.coordinates.vz,
+                    covariance=CoordinateCovariances.from_matrix(covariance),
+                    frame=orbit.coordinates.frame,
+                    origin=orbit.coordinates.origin,
+                ),
+            )
```

Comment:
`mean_state` is unused. I think this could be clearer by not assigning to a name ( like `_,  covariance = mean_and_covariance_from_weighted_samples(...)`). 

You could go further, though, and change `mean_and_covariance_from_weighted_samples`. Instead of returning two values (one of which you won't use), you could just have a function called `weighted_covariance` which computes the covariance from samples using weights. You could also have a `weighted_mean` function, of course. 

Those could be improved again by perhaps making them methods on `VariantOrbits`. Then this code might look like:

```py
for key, orbit, variants in link.iterate():
    key = key.as_py()
    covariance = variants.weighted_covariance()
    
```

which feels pretty nice.

In addition, quivr Tables are "copy-on-write". So you could simplify your big copy to make `orbit_collapsed`:

```py
coords = orbit.coordinates
coords.covariance = CoordinateCovariances.from_matrix(covariance)
orbit.coordinates = coordinates
orbits_list.append(orbit)
```

Putting it all together (and dropping the unused `key`):
```suggestion
        for _, orbit, variants in link.iterate():
            assert len(orbit) == 1

            # Use the weighted sample covariance, and leave all other attributes
            # on the orbit unchanged.
            coords = orbit.coordinates
            coords.covariance = CoordinateCovariances.from_matrix(variants.weighted_covariance())
            orbit.coordinates = coordinates
            orbits_list.append(orbit)
```